newsidh/sidh.go

```diff
@@ -0,0 +1,403 @@
+type SIDHPublicKeyAlice struct {
+	affine_xP   ExtensionFieldElement
+	affine_xQ   ExtensionFieldElement
+	affine_xQmP ExtensionFieldElement
+}
+
+// Bob's secret key.
+type SIDHSecretKeyBob struct {
+	Scalar [SecretKeySize]byte
+}
+
+// Alice's secret key.
+type SIDHSecretKeyAlice struct {
+	Scalar [SecretKeySize]byte
+}
+
+// Read a public key from a byte slice.  The input must be at least 564 bytes long.
+func (e *SIDH) FromBytesBob(input []byte) (pubKey *SIDHPublicKeyBob) {
+	if len(input) < e.publicKeySize {
+		panic("Too short input to SIDH pubkey FromBytes, expected 564 bytes")
+	}
+	pubKey.affine_xP.FromBytes(input[0:e.sharedSecretSize])
+	pubKey.affine_xQ.FromBytes(input[e.sharedSecretSize : 2*e.sharedSecretSize])
+	pubKey.affine_xQmP.FromBytes(input[2*e.sharedSecretSize : 3*e.sharedSecretSize])
+
+	return pubKey
+}
+
+// Write a public key to a byte slice.  The output must be at least 564 bytes long.
+func (e *SIDH) ToBytesBob(pubKey *SIDHPublicKeyBob) (output []byte) {
+	if len(output) < e.publicKeySize {
+		panic("Too short output for SIDH pubkey FromBytes, expected 564 bytes")
+	}
+	pubKey.affine_xP.ToBytes(output[0:e.sharedSecretSize])
+	pubKey.affine_xQ.ToBytes(output[e.sharedSecretSize : 2*e.sharedSecretSize])
+	pubKey.affine_xQmP.ToBytes(output[2*e.sharedSecretSize : 3*e.sharedSecretSize])
+
+	return output
+}
+
+// Read a public key from a byte slice.  The input must be at least 564 bytes long.
+func (e *SIDH) FromBytesAlice(input []byte) (pubKey *SIDHPublicKeyAlice) {
+	if len(input) < e.publicKeySize {
+		panic("Too short input to SIDH pubkey FromBytes, expected 564 bytes")
+	}
+	pubKey.affine_xP.FromBytes(input[0:e.sharedSecretSize])
+	pubKey.affine_xQ.FromBytes(input[e.sharedSecretSize : 2*e.sharedSecretSize])
+	pubKey.affine_xQmP.FromBytes(input[2*e.sharedSecretSize : 3*e.sharedSecretSize])
+
+	return pubKey
+}
+
+// Write a public key to a byte slice.  The output must be at least 564 bytes long.
+func (e *SIDH) ToBytesAlice(pubKey *SIDHPublicKeyAlice) (output []byte) {
+	if len(output) < e.publicKeySize {
+		panic("Too short output for SIDH pubkey FromBytes, expected 564 bytes")
+	}
+	pubKey.affine_xP.ToBytes(output[0:e.sharedSecretSize])
+	pubKey.affine_xQ.ToBytes(output[e.sharedSecretSize : 2*e.sharedSecretSize])
+	pubKey.affine_xQmP.ToBytes(output[2*e.sharedSecretSize : 3*e.sharedSecretSize])
+
+	return output
+}
+
+// Generate a keypair for "Alice".  Note that because this library does not
+// implement SIDH validation, each keypair should be used for at most one
+// shared secret computation.
+func (e *SIDH) GenerateAliceKeypair(rand io.Reader) (publicKey *SIDHPublicKeyAlice, secretKey *SIDHSecretKeyAlice, err error) {
+	publicKey = new(SIDHPublicKeyAlice)
+	secretKey = new(SIDHSecretKeyAlice)
+
+	_, err = io.ReadFull(rand, secretKey.Scalar[:])
+	if err != nil {
+		return nil, nil, err
+	}
+	// Bit-twiddle to ensure scalar is in 2*[0,2^371):
+	secretKey.Scalar[e.secretKeySize-1] = e.maskAliceByte1
```

Comment:
I know that it's a detail but strictly speaking, secret key should be chosen uniformly at random. By applying this mask we favour some numbers over some others